src/core/store/common/tokensByAsset.selectors.ts

```diff
@@ -41,7 +49,11 @@ export const selectDepositTokenOptionsByAsset = createSelector(
           const allowancesMap = userTokensAllowancesMap[address] ?? {};
           return createToken({ tokenData, userTokenData, allowancesMap });
         });
-      return tokens.filter((token) => toBN(token.balance).gt(0) || token.address === mainVaultToken);
+      return tokens.filter(
+        (token) =>
+          isSupportedToken({ assetData, token, mainVaultToken, zapperDisabled, zapType: 'zapInWith' }) &&
+          toBN(token.balance).gt(0)
+      );
```

Comment:
Seems this part is not correct since `mainVaultToken` should be returned even if it does not have balance.

How about something like:
```suggestion
      return tokens.filter(
        (token) => {
           const isZappable = token.supported[assetData.metadata.zapInWith] && toBN(token.balance).gt(0);
           return isZappable || token.address === mainVaultToken;
        }
      );
```

Reply:
I think that works, I've refactored to that other function because at the minute we need to cast and it becomes noisy. What do you think of https://github.com/yearn/yearn-finance-v3/pull/691/commits/58ab9e2b15f782a16eb0ccee76556690f4f62323